File: app/Sources/ICD360SVPN/Models/APIError.swift
// ICD360SVPN — Models/APIError.swift
// MARK: - Typed API error
//
// Anything that can go wrong during a call to vpn-agent is normalised
// into one of these cases. The view layer matches on them to render
// readable feedback (and to distinguish "you need to enroll" from
// "the server is down").

import Foundation

public enum APIError: Error, LocalizedError {
    /// Lower-level transport problem (DNS, TCP, TLS handshake, …).
    case transport(Error)

    /// Server returned a non-HTTPURLResponse — should never happen.
    case invalidResponse

    /// Non-2xx response with a body that we could not parse as
    /// `application/problem+json`.
    case http(status: Int, body: Data?)

    /// RFC 7807 problem+json error returned by the agent.
    case problemDetails(ProblemDetails)

    /// JSON decoding failed.
    case decoding(Error)

    /// No client identity in the keychain — caller needs to enroll.
    case missingIdentity

    /// RFC 7807 Problem Details for HTTP APIs.
    public struct ProblemDetails: Decodable, Sendable {
        public let type: String
        public let title: String
        public let status: Int
        public let detail: String?
    }

    public var errorDescription: String? {
        switch self {
        case .transport(let err):
            return "Network error: \(err.localizedDescription)"
        case .invalidResponse:
            return "The server returned a malformed response."
        case .http(let status, _):
            return "HTTP \(status) from the agent."
        case .problemDetails(let pd):
            if let detail = pd.detail, !detail.isEmpty {
                return "\(pd.title): \(detail)"
            }
            return pd.title
        case .decoding(let err):
            return "Could not decode the agent's response: \(err.localizedDescription)"
        case .missingIdentity:
            return "No saved credentials. Please enroll first."
        }
    }
}

File: app/lib/src/models/api_error.dart
// ICD360SVPN — lib/src/models/api_error.dart
//
// Anything that can go wrong while talking to vpn-agent is normalised
// into one of these. The view layer matches on .kind for special
// cases (e.g. .missingIdentity → bounce back to enrollment).

class ApiError implements Exception {
  ApiError({required this.kind, required this.message, this.statusCode});

  final ApiErrorKind kind;
  final String message;
  final int? statusCode;

  @override
  String toString() => 'ApiError($kind): $message';

  /// Build an [ApiError] from an RFC 7807 application/problem+json
  /// response body.
  factory ApiError.fromProblemDetails(int statusCode, Map<String, dynamic> body) {
    final detail = body['detail'] as String?;
    final title = body['title'] as String?;
    return ApiError(
      kind: ApiErrorKind.problemDetails,
      message: detail?.isNotEmpty == true ? '$title: $detail' : (title ?? 'HTTP $statusCode'),
      statusCode: statusCode,
    );
  }
}

enum ApiErrorKind {
  /// Lower-level transport problem: DNS, TCP, TLS handshake.
  transport,

  /// Server returned a non-2xx response we could not parse as
  /// problem+json.
  http,

  /// RFC 7807 problem+json response.
  problemDetails,

  /// JSON decoding failed on a 2xx body.
  decoding,

  /// No client identity in secure storage — caller must enroll first.
  missingIdentity,
}
